### PR TITLE
Various updates

### PR DESCRIPTION
**Pull request by bmn**
Mostly:
* Leaderboard update; solo variant, etc.
* Alt. component handles on multi-decker bombs
* Module-specific help/manual messages
* Hold the bomb at the start of the round
* Aliases and friendlier inputs for modules/binders
* Extra info provided by the chat bot in certain situations

### CODE REVIEW

File: Assets/Scripts/Commanders/BombBinderCommander.cs [hunk 1]
```diff
@@ -101,6 +101,21 @@ public IEnumerator RespondToCommand(string userNickName, string message, IComman
                     yield return selectCoroutine.Current;
                 }
             }
+            else if (message.StartsWith("select", StringComparison.InvariantCultureIgnoreCase))
```
Comment (ashbash1987) [inline Assets/Scripts/Commanders/BombBinderCommander.cs:104]: _Reminder:_ Add to wiki description on `!binder select <x>` alternative.

File: Assets/Scripts/Commanders/BombBinderCommander.cs [hunk 2]
```diff
@@ -225,8 +240,39 @@ private void MoveUpOnPage()
...
+            if ( (_currentSelectables == null) || (index > _currentSelectables.Length) )
+            {
+                yield break;
+            }
+
+            int i = 0;
+            MonoBehaviour newSelectable = null;
+            for (_currentSelectableIndex = 0; _currentSelectableIndex < _currentSelectables.Length; ++_currentSelectableIndex)
+            {
+                Debug.Log(string.Format("[bmn] Looking for index {0}", _currentSelectableIndex));
```
Comment (ashbash1987) [inline Assets/Scripts/Commanders/BombBinderCommander.cs]: Would ideally prefer log messages with regards to Twitch Plays not have user-specific log tags if possible. Can keep the log message no problem, but just without the `[bmn]` part I guess.

File: Assets/Scripts/Commanders/BombCommander.cs [hunk 3]
```diff
@@ -124,8 +125,7 @@ public IEnumerator RespondToCommand(string userNickName, string message, IComman
 
             responseNotifier.ProcessResponse(CommandResponse.EndNotComplete);
         }
-        else if (message.Equals("time", StringComparison.InvariantCultureIgnoreCase) ||
-                 message.Equals("timestamp", StringComparison.InvariantCultureIgnoreCase) ||
+        else if (message.Equals("timestamp", StringComparison.InvariantCultureIgnoreCase) ||
```
Comment (ashbash1987) [inline Assets/Scripts/Commanders/BombCommander.cs:128]: _Reminder:_ Add to wiki description on `!bomb timestamp`/`!bomb date`

(Also, not part of this pull request, I know, another reminder to add wiki description on `!bomb edgework <x>` and variants for particular faces)

File: Assets/Scripts/Commanders/BombCommander.cs [hunk 4]
```diff
@@ -140,6 +140,24 @@ public IEnumerator RespondToCommand(string userNickName, string message, IComman
 
             responseNotifier.ProcessResponse(CommandResponse.EndNotComplete);
         }
+        else if (message.Equals("help", StringComparison.InvariantCultureIgnoreCase))
```
Comment (ashbash1987) [inline Assets/Scripts/Commanders/BombCommander.cs:143]: _Reminder:_ Add to wiki description on `!bomb help`

File: Assets/Scripts/Commanders/BombCommander.cs [hunk 5]
```diff
@@ -140,6 +140,24 @@ public IEnumerator RespondToCommand(string userNickName, string message, IComman
...
+        {
+            responseNotifier.ProcessResponse(CommandResponse.Start);
+
+            yield return "sendtochat The Bomb: Pick up with !bomb hold. Turn with !bomb turn. Show the edges with !bomb edgework. Show a specific edge with !bomb top. Display the bomb start time with !bomb time. Edges are top, bottom, left and right.";
+
+            responseNotifier.ProcessResponse(CommandResponse.EndNotComplete);
+        }
+        else if (message.Equals("time", StringComparison.InvariantCultureIgnoreCase) ||
```
Comment (ashbash1987) [inline Assets/Scripts/Commanders/BombCommander.cs:151]: _Reminder:_ Add to wiki description on `!bomb time`/`!bomb timer`/`!bomb clock`

File: Assets/Scripts/ComponentSolvers/ComponentSolver.cs [hunk 6]
```diff
@@ -198,6 +209,16 @@ private bool OnPass(object _ignore)
             AwardSolve(_currentUserNickName, _currentResponseNotifier);
         }
 
+        BombCommander._bombSolvedModules++;
+        
+        if (_turnQueued)
+        {
+            Debug.Log("[bmn] Activating queued turn for completed module.");
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/ComponentSolver.cs]: Again, ideally remove the user-specific `[bmn]` from the logging.

File: Assets/Scripts/ComponentSolvers/Vanilla/InvisibleWallsComponentSolver.cs [hunk 7]
```diff
@@ -25,32 +27,18 @@ protected override IEnumerator RespondToCommandInternal(string inputCommand)
...
+        foreach (Match move in Regex.Matches(inputCommand, @"[udlr]", RegexOptions.IgnoreCase))
         {
-            MonoBehaviour button = null;
-
-            if (buttonString.Equals("u", StringComparison.InvariantCultureIgnoreCase) || buttonString.Equals("up", StringComparison.InvariantCultureIgnoreCase))
-            {
-                button = (MonoBehaviour)_buttons[0];
-            }
-            else if (buttonString.Equals("l", StringComparison.InvariantCultureIgnoreCase) || buttonString.Equals("left", StringComparison.InvariantCultureIgnoreCase))
+            Dictionary<string, int> buttonIndex = new Dictionary<string, int>
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/InvisibleWallsComponentSolver.cs]: Since this dictionary is just a static lookup, it'd be better placed as a `static readonly` at the top of the class instead of being potentially re-initialised/re-instanced for each match. The compiler might be clever and realise it is essentially a const-expression, but either way would be better for me coded explicitly as a static lookup.

File: Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs [hunk 8]
```diff
@@ -31,7 +31,8 @@ protected override IEnumerator RespondToCommandInternal(string inputCommand)
         if (buttonNumber >= 1 && buttonNumber <= 4)
         {
             if (commandParts[0].Equals("position", StringComparison.InvariantCultureIgnoreCase) ||
-                commandParts[0].Equals("pos", StringComparison.InvariantCultureIgnoreCase))
+                commandParts[0].Equals("pos", StringComparison.InvariantCultureIgnoreCase) ||
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs:34]: _Reminder:_ Add to wiki description on `!<id> pos <x>`/`!<id> p <x>`

File: Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs [hunk 9]
```diff
@@ -40,7 +41,9 @@ protected override IEnumerator RespondToCommandInternal(string inputCommand)
                 yield return new WaitForSeconds(0.1f);
                 DoInteractionEnd(button);
             }
-            else if (commandParts[0].Equals("label", StringComparison.InvariantCultureIgnoreCase))
+            else if (commandParts[0].Equals("label", StringComparison.InvariantCultureIgnoreCase) ||
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs:44]: _Reminder:_ Add to wiki description on `!<id> lab <x>`/`!<id> l <x>`

File: Assets/Scripts/ComponentSolvers/Vanilla/MorseCodeComponentSolver.cs [hunk 10]
```diff
@@ -25,13 +26,16 @@ protected override IEnumerator RespondToCommandInternal(string inputCommand)
             yield break;
         }
 
-        if (!commandParts[0].Equals("transmit", StringComparison.InvariantCultureIgnoreCase))
+        if (!commandParts[0].Equals("transmit", StringComparison.InvariantCultureIgnoreCase) &&
+            !commandParts[0].Equals("trans", StringComparison.InvariantCultureIgnoreCase) &&
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/MorseCodeComponentSolver.cs:30]: _Reminder:_ Add to wiki description on `!<id> trans <x>`/`!<id> xmit <x>`/`!<id> tx <x>`

File: Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs [hunk 11]
```diff
@@ -17,7 +17,19 @@ public class PasswordComponentSolver : ComponentSolver
 
     protected override IEnumerator RespondToCommandInternal(string inputCommand)
     {
-        if (Regex.IsMatch(inputCommand, @"^[a-zA-Z]{5}$"))
+        if (inputCommand.Equals("claim", StringComparison.InvariantCultureIgnoreCase))
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs:20]: _Reminder:_ Add to wiki description on `!<id> claim` (as special 'non-function' due to the nature of Twitch Plays playing).

File: Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs [hunk 12]
```diff
@@ -17,7 +17,19 @@ public class PasswordComponentSolver : ComponentSolver
...
+        {
+            yield break;
+        }
+        else if (inputCommand.Equals("cycle", StringComparison.InvariantCultureIgnoreCase))
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs:24]: _Reminder:_ Add to wiki description on `!<id> cycle` (non-`<x>` variant).

File: Assets/Scripts/ComponentSolvers/Vanilla/SimonComponentSolver.cs [hunk 13]
```diff
@@ -23,32 +25,18 @@ protected override IEnumerator RespondToCommandInternal(string inputCommand)
...
-            if (buttonString.Equals("r", StringComparison.InvariantCultureIgnoreCase) || buttonString.Equals("red", StringComparison.InvariantCultureIgnoreCase))
-            {
-                button = (MonoBehaviour)_buttons.GetValue(0);
-            }
-            else if (buttonString.Equals("b", StringComparison.InvariantCultureIgnoreCase) || buttonString.Equals("blue", StringComparison.InvariantCultureIgnoreCase))
-            {
-                button = (MonoBehaviour)_buttons.GetValue(1);
-            }
-            else if (buttonString.Equals("g", StringComparison.InvariantCultureIgnoreCase) || buttonString.Equals("green", StringComparison.InvariantCultureIgnoreCase))
+            Dictionary<string, int> buttonIndex = new Dictionary<string, int>
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/SimonComponentSolver.cs]: Same again as previous Simon comment; since this dictionary is just a static lookup, it'd be better placed as a static readonly at the top of the class instead of being potentially re-initialised/re-instanced for each match. The compiler might be clever and realise it is essentially a const-expression, but either way would be better for me coded explicitly as a static lookup.

File: Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs [hunk 14]
```diff
@@ -18,15 +18,17 @@ public class WireSequenceComponentSolver : ComponentSolver
 
     protected override IEnumerator RespondToCommandInternal(string inputCommand)
     {
-        if (inputCommand.Equals("up", StringComparison.InvariantCultureIgnoreCase))
+        if (inputCommand.Equals("up", StringComparison.InvariantCultureIgnoreCase) ||
+            inputCommand.Equals("u", StringComparison.InvariantCultureIgnoreCase))
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs:22]: _Reminder:_ Add to wiki description on `!<id> u`.

File: Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs [hunk 15]
```diff
@@ -18,15 +18,17 @@ public class WireSequenceComponentSolver : ComponentSolver
...
         {
             yield return "up";
 
             DoInteractionStart(_upButton);
             yield return new WaitForSeconds(0.1f);
             DoInteractionEnd(_upButton);
         }
-        else if (inputCommand.Equals("down", StringComparison.InvariantCultureIgnoreCase))
+        else if (inputCommand.Equals("down", StringComparison.InvariantCultureIgnoreCase) ||
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs:30]: _Reminder:_ Add to wiki description on `!<id> d`.

File: Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs [hunk 16]
```diff
@@ -48,6 +50,25 @@ protected override IEnumerator RespondToCommandInternal(string inputCommand)
 
             foreach (string wireIndexString in sequence)
             {
+                if (wireIndexString.Equals("up", StringComparison.InvariantCultureIgnoreCase) ||
```
Comment (ashbash1987) [inline Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs:53]: _Reminder:_ Add to wiki description on chaining cuts with movement.

File: Assets/Scripts/Leaderboard.cs [hunk 17]
```diff
@@ -144,7 +309,129 @@ private static int CompareScores(LeaderboardEntry lhs, LeaderboardEntry rhs)
...
+            };
+            string jsonOutput = JsonConvert.SerializeObject(stats, new JsonSerializerSettings()
+            {
+                ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore
+            });
+            File.WriteAllText(path, jsonOutput);
+        }
+        catch (FileNotFoundException)
+        {
+            Debug.LogWarningFormat("[bmn] File {0} was not found.", path);
```
Comment (ashbash1987) [inline Assets/Scripts/Leaderboard.cs]: Again, ideally remove the user-specific `[bmn]` from the logging.

File: Assets/Scripts/UI/TwitchComponentHandle.cs [hunk 18]
```diff
@@ -98,45 +100,57 @@ private void Start()
...
         {
             return;
         }
 
-        string internalCommand = match.Groups[1].Value;
+        string targetModule = match.Groups[1].Value;
+        string internalCommand = match.Groups[2].Value;
         
         string messageOut = null;
         if (internalCommand.Equals("help", StringComparison.InvariantCultureIgnoreCase)) {
```
Comment (ashbash1987) [inline Assets/Scripts/UI/TwitchComponentHandle.cs:156]: Not part of this pull request, but adding reminder for the `!<id> help` and `!<id> manual` commands.

File: Assets/Scripts/UI/TwitchComponentHandle.cs [hunk 19]
```diff
@@ -157,6 +171,11 @@ public void OnMessageReceived(string userNickName, string userColor, string text
             }
             messageOut = string.Format("{0}: https://ktane.timwi.de/HTML/{1}.html", manualText, Uri.EscapeDataString(manualText));
         }
+        else if (Regex.IsMatch(internalCommand, "^(bomb|queue) (turn( a?round)?|flip|spin)$", RegexOptions.IgnoreCase))
```
Comment (ashbash1987) [inline Assets/Scripts/UI/TwitchComponentHandle.cs:174]: _Reminder:_ Add to wiki description on `!<id> bomb turn round` (and variants) for queuing a bomb turn on module solve.

File: Assets/Scripts/UI/TwitchLeaderboard.cs [hunk 20]
```diff
@@ -31,50 +27,74 @@ private void Start()
             return;
         }
 
-        float delay = 0.6f;
-        int index = 0;
+        if (leaderboard.Success)
+        {
+            promptTransform.FindChild("Retry").gameObject.SetActive(false);
```
Comment (ashbash1987) [inline Assets/Scripts/UI/TwitchLeaderboard.cs]: This looks nasty. Could you not just have a reference direct to that Retry GameObject (rather than Transform) assigned in the inspector, and then it's just a nice simple `RetryText.SetActive(false)`? In fact, all of the uses of `promptTransform` I see are in fact GameObject uses instead.

File: Assets/Scripts/UI/TwitchLeaderboard.cs [hunk 21]
```diff
@@ -31,50 +27,74 @@ private void Start()
...
+            mainTable = Instantiate<TwitchLeaderboardTable>(twitchLeaderboardTablePrefab);
+            mainTable.leaderboard = leaderboard;
+        }
 
-            _instancedRows.Add(row);
+        if (leaderboard.SoloCount > 0)
+        {
+            soloTable = Instantiate<TwitchLeaderboardTableSolo>(twitchLeaderboardTableSoloPrefab);
+            soloTable.leaderboard = leaderboard;
+            transform.FindChild("UI").FindChild("Mask").gameObject.SetActive(true);
```
Comment (ashbash1987) [inline Assets/Scripts/UI/TwitchLeaderboard.cs]: This also looks decidedly nasty. And I'm not particularly sure what it's actually trying to achieve. I can see there's a mask involved in some way, but don't actually know for what purpose currently.

File: Assets/Scripts/UI/TwitchLeaderboard.cs [hunk 22]
```diff
@@ -31,50 +27,74 @@ private void Start()
...
+            soloTable.transform.SetParent(prioritiseSolo ? mainTableTransform : altTableTransform, false);
+            mainTable.transform.SetParent(prioritiseSolo ? altTableTransform : mainTableTransform, false);
         }
         else
         {
-            totalSolveRate = ((float)solveCount) / strikeCount;
+            if ((leaderboard.Count - statsTable.entriesLess) > mainTable.maximumRowCount)
+            {
+                statsTable.transform.SetParent(altTableTransform, false);
+                transform.FindChild("UI").FindChild("Mask").gameObject.SetActive(true);
```
Comment (ashbash1987) [inline Assets/Scripts/UI/TwitchLeaderboard.cs]: Again, quite nasty-looking as before.

**Review by ashbash1987 — COMMENTED** [review]
Ok, I finally got to reviewing this pull request! I'm pretty much happy to accept it, but there's just a few niggles that could probably be sorted out quite quickly.

For the comments where I've started it with _Reminder:_, that's a reminder for either you or me to update the wiki page with the new commands available. So don't take those comments as *I must do these* requests :P

**Comment by bmn** [comment]
No worries, will sort out what I can :) I'm definitely a beginner here - making a mess of your mod is the first thing I've done in either C# or Unity - so the criticisms are probably very valid! Will reply to specifics in detail.

File: Assets/Scripts/Commanders/BombBinderCommander.cs (hunk 2 above)
Comment (bmn) [reply Assets/Scripts/Commanders/BombBinderCommander.cs]: The user-specific parts of these log messages were something I intended to take out before PRing but apparently missed some. I'll go over it properly and update.

File: Assets/Scripts/ComponentSolvers/Vanilla/InvisibleWallsComponentSolver.cs (hunk 7 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/InvisibleWallsComponentSolver.cs]: Agreed, will move it.

File: Assets/Scripts/ComponentSolvers/Vanilla/SimonComponentSolver.cs (hunk 13 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/SimonComponentSolver.cs]: Agreed.

File: Assets/Scripts/UI/TwitchLeaderboard.cs (hunk 20 above)
Comment (bmn) [reply Assets/Scripts/UI/TwitchLeaderboard.cs]: Agreed on this, not sure why I went that route in the first place.

File: Assets/Scripts/UI/TwitchLeaderboard.cs (hunk 21 above)
Comment (bmn) [reply Assets/Scripts/UI/TwitchLeaderboard.cs]: The "mask" is a black background and gradient on the left side of the prefab (TwitchLeaderboard/UI/Mask), which is just there to cover over the blue results binder when there's content on the left side. The reason for activating it in code is that, if there have been no solo clears and less than 16 players, the leaderboard uses the same right-side-only layout as the current TP version.

I'm a bit uncomfortable with this, as I'm pretty sure using a black box isn't the right way to cover something up. I don't have the knowledge of Unity currently to do this in a more appropriate manner, so would appreciate some input from someone with more experience.

File: Assets/Scripts/UI/TwitchLeaderboard.cs (hunk 22 above)
Comment (bmn) [reply Assets/Scripts/UI/TwitchLeaderboard.cs]: Agreed.

File: Assets/Scripts/UI/TwitchLeaderboard.cs (hunk 21 above)
Comment (ashbash1987) [reply Assets/Scripts/UI/TwitchLeaderboard.cs]: Ah ok. Covering it up with a black box is fine really (not overly concerned by the amount of overdraw concerned).

File: Assets/Scripts/Commanders/BombBinderCommander.cs (hunk 1 above)
Comment (bmn) [reply Assets/Scripts/Commanders/BombBinderCommander.cs:104]: Done.

File: Assets/Scripts/Commanders/BombCommander.cs (hunk 3 above)
Comment (bmn) [reply Assets/Scripts/Commanders/BombCommander.cs:128]: Done (both parts).

File: Assets/Scripts/Commanders/BombCommander.cs (hunk 4 above)
Comment (bmn) [reply Assets/Scripts/Commanders/BombCommander.cs:143]: Done.

File: Assets/Scripts/Commanders/BombCommander.cs (hunk 5 above)
Comment (bmn) [reply Assets/Scripts/Commanders/BombCommander.cs:151]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs (hunk 8 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs:34]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs (hunk 9 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/MemoryComponentSolver.cs:44]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/MorseCodeComponentSolver.cs (hunk 10 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/MorseCodeComponentSolver.cs:30]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs (hunk 11 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs:20]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs (hunk 12 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/PasswordComponentSolver.cs:24]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs (hunk 14 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs:22]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs (hunk 15 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs:30]: Done.

File: Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs (hunk 16 above)
Comment (bmn) [reply Assets/Scripts/ComponentSolvers/Vanilla/WireSequenceComponentSolver.cs:53]: Done.

File: Assets/Scripts/UI/TwitchComponentHandle.cs (hunk 19 above)
Comment (bmn) [reply Assets/Scripts/UI/TwitchComponentHandle.cs:174]: Done.

File: Assets/Scripts/UI/TwitchComponentHandle.cs (hunk 18 above)
Comment (bmn) [reply Assets/Scripts/UI/TwitchComponentHandle.cs:156]: Done.

**Comment by bmn** [comment]
Should hopefully be all resolved now.

**Review by ashbash1987 — APPROVED** [review]
Ok, here we go!
